### DISCUSSION TITLE
How to declare a build dependency that is a cask?

### DISCUSSION THREAD

**Discussion by xiao-pp** [Homebrew/discussions#1075] [Everyday usage]
Hi,

Our project needs GraalVM to build the bottle binary. However we do not know how to declare it.
```ruby
# this gives syntax error
depends_on cask: "graalvm/tap/graalvm-ce-lts-java11" => :build

# this says the dep cannot be found however suggests there is a cask for it
depends_on "graalvm/tap/graalvm-ce-lts-java11" => :build
```

By installing GraalVM with brew manually our binary is built successfully. However we'd like to declare it so it's managed.

Thanks!

**Comment by SMillerDev** [comment]
The best way would be to get a GraalVM formula in homebrew core. I'm not sure if there really is another way.

**Comment by gromgit** [comment]
Sounds like you could use a custom requirement. You can see [my FUSE tap](https://github.com/gromgit/homebrew-fuse/) for the MacFUSE requirement I pulled out of the core in `require/macfuse.rb`, and how it's used in `Formula/encfs.rb`.

I don't know if you can actually trigger an install from a requirement, but at least you can let users know what to do.

**Comment by anentropic** [comment]
I have exactly the same issue with the same dependency

The docs say _"A Symbol (e.g. `:xcode`) specifies a Requirement which can be fulfilled by one or more formulae, **casks** or other system-wide installed software"_

...but `:"graalvm/tap/graalvm-ce-java11"` doesn't work, I guess because it's a tap path?

@gromgit thanks for the tip re implementing a custom `Requirement`

**Reply by gromgit** [reply]
> ...but `:"graalvm/tap/graalvm-ce-java11"` doesn't work, I guess because it's a tap path?

It doesn't work because a Symbol is an _identifier_, not an string with path semantics: https://www.rubyguides.com/2018/02/ruby-symbols/

In the case of dependency symbols, a simple `grep` suggests that they're mapped to specific Requirement classes here:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dependency_collector.rb#L119-L130

**Reply by anentropic** [reply]
my mistake for reading the docs instead of looking in the src...

**Reply by gromgit** [reply]
Well, if you dived right into the source without first reading the docs, you'd be lost in short order. The Formula Cookbook is intended to get you started in understanding, modifying and creating formulae, not a comprehensive reference like the [Formula RubyDoc](https://rubydoc.brew.sh/Formula).

**Reply by anentropic** [reply]
yes, what's missing are good docs

the cookbook is one long page that lacks structure and sufficient detail, and the RubyDoc pages are a big dump of stuff auto-generated from the src code without context

**Reply by SMillerDev** [reply]
> yes, what's missing are good docs

Pull requests are welcome. But we're a package manager, our focus is mostly on packaging software.
